src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs

```diff
@@ -23,6 +23,23 @@ public sealed class FormatHex : PSCmdlet
     {
         private const int BUFFERSIZE = 16;
 
+        /// <summary>
+        /// For cases where a homogenous collection of bytes or other items are directly piped in, we collect all the
+        /// bytes in a List&lt;byte&gt; and then output the formatted result all at once in EndProcessing().
```

Comment:
Scenario: convert to hex whole _large_ file and write to text file for review. In the case the collecting can eat all memory.

Reply:
Would this not be essentially similar to any cmdlet that might be operating on large files and collecting the entire contents? e.g., Sort-Object? I guess there's a bit extra overhead because of generating the hex data, but the majority of the load with a large file should be basically the same as that.

I'll look for something to test it on tomorrow, I suppose. 😄

Reply:
My thoughts is that if we can not collect we should not collect.
Or maybe convert by blocks (by lines or by 512 bytes)? Some kind of lookahead...

Reply:
By blocks doesn't sound like a bad idea overall... 🤔

Maybe a combined behaviour here would make sense? Something along these lines:

* Strings are output per-line (every input string, essentially)
* Bytes are collected in blocks of up to 512 (or more?) and then output.
* Other items are collected in blobs that sum to at most 512 bytes and then output

Between each, the list can be cleared. When handling mixed input types, perhaps it makes the most sense to collate similar types (e.g., if I pipe in twelve bytes and then a string, output the twelve bytes _together_ and then the string separately)?

What I would _like_ to do is modify the formatted output such that if I output X bytes as a blob, and _then_ output the next 512 bytes, the line-counter is not reset, _and_ the headers do not show again. At least, I think that might improve the overall display?

Reply:
I believe by default output should be "by item". To address second scenario we could add new parameter
```c#
 fhx -ByBlock 17
fhx -ByBlock 256
``

Reply:
That would have to assume uniform input, though, and doesn't address any of the mixed input scenarios? It would be very strange to have such a parameter only affect one kind of input, and it would be even stranger to try to apply such a parameter to the byte-width of other types, surely?

Reply:
>  any of the mixed input scenarios

We could enable fhx -ByItem -ByBlock or introduce an enum - fhx -By (Item, Block, ItemThenBlock)

Reply:
I'm thinking perhaps we follow Format-Table's precedent albeit in a slightly different vein with `-GroupBy [ Item | Block | Continuous | Mixed ]`

Or something like that, anyhow. 😄 I think `Mixed` (block similar numeric value-types that come in as a continuous stream of items, and then output everything else on a per-item basis).

I would still like to modify the default formatted output to allow me to output even a continuous block line-by line as it comes in (so for lines after the first, the headers are skipped) with a different overload. Do you have any thoughts on the best way to accomplish that? That would allow us to render a massive amount of bytes as a block without actually necessitating them being collected for any more than a single line as they would be output and then cleared from the cache as they're rendered.

Reply:
I think we need more feedback the formatting and parameter before discussing of implementation details.
@SteveL-MSFT @mklement0 Have you any thoughts?

Reply:
@SteveL-MSFT need your input/suggestions on formatting.

Reply:
ping @SteveL-MSFT

Reply:
Sorry, missed this til now.  The concern is a single large file being read into memory and then processed?  I would think most users would use `Get-Content foo.txt | Format-Hex` in which case, `Get-Content` sends out line by line to the pipeline so at no point does the entire file get read into memory (unless -Raw is used).  If a byte-array is being used, then collecting up to a line's worth and then outputting seems ok.  Any other object should be per object unless I'm missing something here.

Reply:
I guess the question becomes -- now that Format-Hex is capable of processing and using a variety of possible inputs (really basically anything it can marshal to bytes, thanks to Ilya's implementation of Jason's initial code):

1. do we really want it to clump similar items together?
1. or do we want it to separate out individual items?
1. What if it's mixed input? You give it a string, a handful of characters, and some random integers all in the same pipeline?
    a. Should it group? 
    a. Should it output each item individually?
    a. Should this behaviour be configurable?
    a. If so, what would be a sensible default?

Of course, I have my ideas laid out already in the code and explained above (though I haven't yet made the behaviour configurable).

Also, I'm of the opinion we should provide a way for the formatter to render multiple output entries as a single "block" when it may be sensible, rather than force every output object to display its own header.

Reply:
@iSazonov's worry with the current implementation can be alleviated if and only if that last is made available to the cmdlet, otherwise we run the risk of making the formatting unreadable.

Reply:
@vexx32 thanks for the clarification.  Now your grouping comment makes sense to me.  I guess the question is what is the use case where someone sends different types of objects in the same pipeline to Format-Hex?  My default feeling is that grouping it per different object type makes sense.  Since this is strictly formatting, it's not considered a breaking change, so I would say go with something reasonable and not complex and we can always adjust based on user feedback.

Reply:
Sounds good to me!

I'll look at adjusting the behaviour for strings so that we can side-step the issue where you pipe a large file into the cmdlet.

Apart from that I think this is ready! I will make the change today or tomorrow.

Reply:
Then please update the PR description with output samples.

Reply:
@iSazonov added an example showing a possible outcome where individual strings end up piped in

I don't think it makes sense to just cram them all together based on byte count, as we're already missing the line endings in the Get-Content case, and we can't assume input strings come from get-content. 

In that kind of case, I think it might make the most sense to simply list the index of the item rather than the byte offset?

Reply:
@vexx32 Just want to make sure one of my primary use scenarios works well.  What do you get if you do this:
```
-532265403 | format-hex
```
I'm hoping it would be something easily/visibly constructible back to `0xe0464645`.

Reply:
You would get what you currently get in 7-preview1:
```code
                       00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
00000000000000000000   45 46 46 E0                                      EFFà
```

Reply:
@iSazonov having had this time to think on it, I think you're right in that for most objects rendering a table for every pipeline input object is the most sensible.

However, in the case of raw bytes being piped in, I think it's most sensible to group them into a single output object.

So, you'd render individual blocks:

- for strings
- for arbitrary value types

But you'd render a single block:

- for bytes

Having the default grouping behaviour clump bytes I feel makes the most sense; I'm not sure I can really see value in grouping most other object types together?

Reply:
@vexx32 OK, maybe in future PR we could add support for a `WordWidth` parameter or something like that.  If anything, that also comes in handy when viewing Unicode as a double-byte sequence e.g.:
```
                       00   02   04   06   08   0A   0C   0E 
00000000000000000000   0048 0065 006C 006C 006F                                   Hello
```

Reply:
> But you'd render a single block:
for bytes

Maybe chars and other primitive types too. We need to see output samples to make conclusion.

Reply:
Chars I'd agree with, but most other primitives don't seem to make a lot of sense to be grouped.

I'll scrape together some output samples shortly so we can see what having them all grouped into a single block would look like. 🙂

Reply:
I think 16-bit, 32-bit and 64-bit values make a lot of sense to be grouped together (primarily because of Intel's little endianness.  For instance, this `E0464545` is easier to read (and reflects the actual hex value) than `45 46 46 E0`.

Reply:
@rkeithhill Format-Hex is displaying the hex bytes as they are for these values -- the bytes and the order they're in is coming directly from the CLR, it's not re-ordering them. You can see the code as it already stands in the master branch here, it was @lzybkr's concept that @iSazonov implemented a few months(?) ago now.

Apologies for not really being clear. As a brief example, when I say output would be "grouped" vs "by item", this is what I mean:

**Grouped**
```
                       00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
00000000000000000000   01 00 00 00 02 00 00 00 03 00 00 00 04 00 00 00  ................
00000000000000000010   05 00 00 00 06 00 00 00                          ........
```

**By item**
```
                       00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
00000000000000000000   01 00 00 00                                      ....
----------------------------------------------------------------------------
00000000000000000000   02 00 00 00                                      ....
----------------------------------------------------------------------------
00000000000000000000   03 00 00 00                                      ....
----------------------------------------------------------------------------
00000000000000000000   04 00 00 00                                      ....
----------------------------------------------------------------------------
00000000000000000000   05 00 00 00                                      ....
----------------------------------------------------------------------------
00000000000000000000   06 00 00 00                                      ....
```

Reply:
@iSazonov further output samples if we opt to group all primitives. I'm using `Format-Hex -InputObject` for the mockup, but the proposal here is that piping the input to the cmdlet would result in the same behaviour.



**Byte**

I think this one's essentially a no-brainer. It seems weirder _not_ to.
```
$a = [byte[]] ( [byte]::MinValue..[byte]::MaxValue | Get-Random -Count 32 )
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61311643-c611e180-a7c4-11e9-91de-bf694e7f37d3.png)
```
$a = 1..64 -as [byte[]]
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61312352-3ec56d80-a7c6-11e9-89a3-ea794d069713.png)

**Char**
```
$a = [char[]] ( [char]::MinValue..[char]::MaxValue | Get-Random -Count 32 )
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61311445-5ac80f80-a7c4-11e9-9a17-a9e24fb1a2da.png)
```
$a = 1..64 -as [char[]]
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61311934-649e4280-a7c5-11e9-9ea6-6f515be7ce47.png)

**Int16**
```
$a = [short[]] ( [short]::MinValue..[short]::MaxValue | Get-Random -Count 32 )
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61311676-db870b80-a7c4-11e9-8393-a58ea68e7e52.png)
```
$a = 1..64 -as [short[]]
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61311989-839cd480-a7c5-11e9-8c28-76780de9ecf2.png)

**Int32**
```
$a = [int[]] @(Get-Random -Minimum ([int]::MinValue) -Maximum ([int]::MaxValue) -Count 16)
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61310974-57805400-a7c3-11e9-85eb-83969bd7bd05.png)
```
$a = 1..64 -as [int[]]
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61312041-9d3e1c00-a7c5-11e9-8c90-0b7fa28fe974.png)

**Int64**
```
$a = [long[]] @(Get-Random -Minimum ([long]::MinValue) -Maximum ([long]::MaxValue) -Count 16)
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61311017-7252c880-a7c3-11e9-8842-6a194c6c2ba3.png)
```
$a = 1..32 -as [long[]]
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61312116-c3fc5280-a7c5-11e9-97c1-d2c704dd83dc.png)

**Double**
```
$a = [double[]] @(Get-Random -Minimum ([double]::MinValue) -Maximum ([double]::MaxValue) -Count 16)
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61312253-09208480-a7c6-11e9-8db1-9cbcd034a752.png)
```
$a = 1..32 -as [double[]]
Format-Hex -InputObject $a
```
![image](https://user-images.githubusercontent.com/32407840/61312231-fd34c280-a7c5-11e9-82e8-dbda2caf2253.png)

Note: I tried to do this with `[decimal]` as well, but the cmdlet isn't currently capable(?) of transforming `[decimal]` values to hexadecimal. Not sure why.

Did I miss anything? :blush:

Reply:
The samples is good for me.
I don't remember about decimal. We could fix this in follow PR.